---
title: "PG9_Lung_Microbiome_Tutorial"
author: "Christopher Brown"
date: "January 22, 2020"
output: pdf_document
---
This report was built with R version `r getRversion()`. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# I. Introduction
Welcome, introduction goes here.

```{r libraries, message = FALSE}
# This section contains typical libraries used in these analyses

# I. universal data wrangling and visualization
library(tidyverse)      # ggplot, dplyr, purrr, forcats, stringr... and other packages. See https://tidyverse.tidyverse.org/

# II. formatting, plotting
library(scales)         # used in formatting relative abundance % labels; dollar_format()
library(RColorBrewer)   # for generating custom palettes

# III. machine learning & analysis packages
library(vegan)          # decostand(), metaMDS(), scores(), rda(), ordispider(), adonis()
```

# II. Data Preparation

```{r load_16S_rawdata}
# For ease of use, place the processed 16S data files of interest into the same directory as your current RStudio project
# Best practices: keep a separate project for each experiment or combination of experiments
# include a data folder within the project's directory to keep source files organized; also somewhat mitigates the risk of modifying them inadvertently

# I. Read in .shared (counts of sequences assigned to OTU #s)
# Includes basic processing with selection of OTUs > 0.1 % of the population
otu_raw <- read.table("data/heterogeneity_shared.txt", row.names = 2, header = T)
otu_trim <- otu_raw[, -c(1:2)]
otu_trim <- as.matrix(otu_trim)
otu_tmp <- vegan::decostand(otu_trim, "total") * 100
otu_trim[which(otu_tmp < 0.1)] <- 0
otu_good <- otu_trim[, which(colSums(otu_trim) > 0)]

# Trim _S from end of rownames(otu_good); note that it may be useful to keep these well identifiers in some analyses
rownames(otu_good) <- str_remove(rownames(otu_good), "_S\\d+")

# It's a good idea to inspect the data regularly with tools such as dim(), str(), head(), tail(), summary()... This will help in catching errors earlier
dim(otu_good) # dim(otu_good)[1] observations (samples) x dim(otu_good)[2] features (OTUs above .1% population threshold)
otu_good[1:20, 1:20] # use other means of showing a glimpse of the data?


# II. Read in cons.taxonomy (for each OTU: taxonomic levels down to Genus, if possible)
# Note that OTUs are in rows here rather than columns
# 2 imports are required due to the combination of 2 different separators  
tax1 <- read.table("data/heterogeneity_0.03.cons.taxonomy.txt", sep = "\t", row.names = 1, header = T, colClasses = c("character", "numeric", "character"))
tax2 <- read.table("data/heterogeneity_0.03.cons.taxonomy.txt", sep = ";", skip = 1, col.names = c("", "Phylum", "Class", "Order", "Family", "Genus", "")) %>% 
  dplyr::select(-c(1, 7)) %>% 
  purrr::map_df(stringr::str_replace, "\\(.*.\\)", "") %>%
  purrr::map_df(as.factor)
otu_taxonomy <- data.frame(OTU = rownames(tax1), Size = tax1[, 1], tax2, stringsAsFactors = FALSE)
rownames(otu_taxonomy) <- rownames(tax1)

# trim otu_taxonomy based on OTUs in otu_good
otu_good_taxonomy <- otu_taxonomy[otu_taxonomy$OTU %in% colnames(otu_good), ]
rownames(otu_good_taxonomy) <- intersect(rownames(otu_taxonomy), colnames(otu_good))
otu_good_taxonomy <- as.data.frame(otu_good_taxonomy)

dim(otu_good_taxonomy)
otu_good_taxonomy[1:20,]
```

```{r metadata_wrangling, include = FALSE}
# Load in metadata/environmental data, or create it

# Examples of metadata creation using regular expressions (regex) in a tidyverse pipeline
# Note that extracting metadata from sample names is much easier if naming conventions are consistent in terms of spelling and placement
otu_df <- data.frame(decostand(otu_good, "total") * 100, Sample_name = row.names(otu_good), stringsAsFactors = FALSE) %>%
  # Create Specimen_ctrl from first string of characters prior to first "_"
  mutate(Specimen_ctrl = factor(case_when(str_detect(Sample_name, "^AE_") ~ "AE",
                                       str_detect(Sample_name, "^Blank\\d_") ~ "Empty",
                                       str_detect(Sample_name, "^Heparin_") ~ "Heparin",
                                       str_detect(Sample_name, "^HomogCtrl_") ~ "HomogCtrl",
                                       str_detect(Sample_name, "^IsoCtrl_") ~ "IsoCtrl",
                                       str_detect(Sample_name, "^LPS_") ~ "LPS",
                                       str_detect(Sample_name, "^PerfusionPBS_") ~ "PBS",
                                       str_detect(Sample_name, "^WATERD|^waterC") ~ "WaterNeg",
                                       str_detect(Sample_name, "^mock[CD]") ~ "Mock",
                                       str_detect(Sample_name, "^[:alnum:]+") ~ str_extract(Sample_name, "^[:alnum:]+")), # for experimental groups, place after other steps because this regex would fill in values for controls (may or may not be ideal)
                                levels = c("CR1", "CR2", "JAX1", "JAX2", "AE", "Empty", "Heparin", "HomogCtrl", "IsoCtrl", "LPS", "PBS", "WaterNeg", "Mock")
                )
         ) %>%
  # Create Organ
  mutate(Organ = factor(case_when(str_detect(Sample_name, "_CECUM_") ~ "Cecum",
                                  str_detect(Sample_name, "_LUNG_HOMOG") ~ "Lung",
                                  str_detect(Sample_name, "_NASAL_") ~ "Nasal",
                                  str_detect(Sample_name, "_TONGUE_") ~ "Tongue"),
                        levels = c("Nasal", "Tongue", "Lung", "Cecum")
                ) # sum(is.na(otu_df$Organ)) returns 32 (as expected, controls have NA in this column)
         ) %>%
  # Create Mouse
  mutate(Mouse = as.numeric(str_extract(Sample_name, "(?<=(CR\\d|JAX\\d)_)\\d+") # numbers correspond to mice if a CR or JAX specimen
              ) # sum(is.na(otu_df$Mouse)) returns 32 (as expected, the controls have NA in this column)
         ) %>%
  # metadata columns at front, followed by all of the count data
  dplyr::select(Sample_name:Mouse, everything())
  # note that the data is in wide format
  
otu_df$Specimen_ctrl %>% table() %>% sum() # 192: all samples/rows are accounted for
otu_df$Organ %>% summary() %>% sum() # ditto; 32 NAs for organs
```

# III. Analysis of Controls
[Descriptive text here]

```{r analyze_ae}
# 8 samples - perhaps ideal for use as an example of a control RA plot
# don't include geometric mean
ae_df <- filter(otu_df, Specimen_ctrl == "AE")

# determine order of OTUs based on mean within selected specimens
otu_order <- names(sort(colMeans(ae_df[, stringr::str_detect(colnames(ae_df), "Otu")]), decreasing = TRUE))

# create mean sample
sample_names <- ae_df$Sample_name
temp <- dplyr::select(ae_df, otu_order[1:20])
temp <- rbind(temp, apply(dplyr::select(ae_df, otu_order[1:20]), 2, mean))
sample_names <- append(sample_names, "Arith_Mean")
ae_df <- cbind(Sample_name = forcats::fct_inorder(sample_names), temp)
ae_gg <- tidyr::gather(ae_df, key = OTU, value = Percentage, -stringr::str_which(colnames(ae_df), "Otu", negate = TRUE), factor_key = TRUE)
levels(ae_gg$OTU) <- forcats::fct_inorder(paste0(otu_good_taxonomy[levels(ae_gg$OTU), ]$Genus, " (", levels(ae_gg$OTU), ")"))
```

```{r plot_ae, fig.width = 12, fig.height = 12}
# Encourage participants to run these by commenting out a layer, altering 1 parameter to develop a deeper understanding of how to build up a ggplot object

ggplot(data = ae_gg, aes(x = OTU, y = Percentage, fill = Sample_name)) + 
  geom_col() + 
  facet_grid(Sample_name ~ .) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "top") +
  scale_y_continuous(labels = scales::dollar_format(suffix = "%", prefix = "")) +
  labs(title = "Cross-Comparison of AE Control Replicates", x = "OTU", y = "% Relative Abundance")
```

\newpage
# IV. PCA  
[Descriptive text here]

```{r pal_prep, fig.width=8, include=F}
# for PCA ordinations; choose an 9 that covers the maximum number of factor levels (conditions) within a variable of interest
pcapal <- brewer.pal(9,"Set1")
```

```{r pca_prep}
# example: filter out controls
otu_good_spc <- dplyr::filter(otu_df, !(is.na(Organ))) %>%
  droplevels() # remove levels that are no longer present due to filtering

# PCA of all data (non-control)
otu_good_spc_hel <- decostand(dplyr::select(otu_good_spc, contains("Otu")), "hellinger") # use OTU columns only
otu_good_spc_pca <- rda(otu_good2_hel)

(summary(otu_good_spc_pca)) # 0.1639 0.10222 0.08503 proportion explained by first 3 axes

set.seed(8947) # permanova
adonis(otu_good_spc_hel~otu_good_spc$Specimen_ctrl, method="euclidean", permutations = 10000) # 9.999e-05 ***
```

```{r spider_pca, fig.height=10, fig.width=10}
plot(otu_good_spc_pca, type = "n", font = 2, font.lab = 2, xlab = "PC1 (16.39% Explained)", ylab = "PC2 (10.22% Explained)", main = "PCA of Specimen Origin, colored by Organ", display = "sites")
points(otu_good_spc_pca, pch = 19, col = pcapal[as.numeric(otu_good_spc$Organ)])
vegan::ordispider(otu_good_spc_pca, otu_good_spc$Specimen_ctrl, label = TRUE)
legend("topright", levels(otu_good_spc$Organ), pch = 19, col = pcapal, title = "Organ")
legend("bottomright", legend = "Permanova: Difference between Specimen Origin: 9.999e-05") # Should signifiance test be include in the plot?
```
